### PR TITLE
Feature : Multiple ACS URLs for SAML SP / IdP initiated SLO

### PR DESCRIPTION
**Pull request by thanujalk**

### CODE REVIEW

File: components/carbon-authenticators/saml2-sso-authenticator/org.wso2.carbon.identity.authenticator.saml2.sso.ui/pom.xml [hunk 1]
```diff
@@ -115,7 +115,9 @@
                             org.wso2.carbon.identity.authenticator.saml2.sso.stub;
                             version="${carbon.identity.package.import.version.range}",
                             org.wso2.carbon.identity.authenticator.saml2.sso.stub.types;
-                            version="${carbon.identity.package.import.version.range}"
+                            version="${carbon.identity.package.import.version.range}",
+                            org.wso2.carbon.identity.application.authentication.framework.model.xsd;
```
Comment (hpmtissera) [inline components/carbon-authenticators/saml2-sso-authenticator/org.wso2.carbon.identity.authenticator.saml2.sso.ui/pom.xml]: can we put version here? and why resolution optional

File: components/sso-saml/org.wso2.carbon.identity.sso.saml/src/main/java/org/wso2/carbon/identity/sso/saml/dto/SAMLSSOServiceProviderDTO.java [hunk 2]
```diff
@@ -246,4 +251,48 @@ public boolean isDoValidateSignatureInRequests() {
...
+    public boolean isIdPInitSLOEnabled() {
+        return idPInitSLOEnabled;
+    }
+
+    public void setIdPInitSLOEnabled(boolean idPInitSLOEnabled) {
+        this.idPInitSLOEnabled = idPInitSLOEnabled;
+    }
+
+    public String[] getIdpInitSLOReturnToURLs() {
+        return idpInitSLOReturnToURLs;
```
Comment (hpmtissera) [inline components/sso-saml/org.wso2.carbon.identity.sso.saml/src/main/java/org/wso2/carbon/identity/sso/saml/dto/SAMLSSOServiceProviderDTO.java]: return a clone

File: components/sso-saml/org.wso2.carbon.identity.sso.saml/src/main/java/org/wso2/carbon/identity/sso/saml/dto/SAMLSSOServiceProviderDTO.java [hunk 3]
```diff
@@ -246,4 +251,48 @@ public boolean isDoValidateSignatureInRequests() {
...
+    }
+
+    public void setIdpInitSLOReturnToURLs(String[] idpInitSLOReturnToURLs) {
+        this.idpInitSLOReturnToURLs = idpInitSLOReturnToURLs;
```
Comment (hpmtissera) [inline components/sso-saml/org.wso2.carbon.identity.sso.saml/src/main/java/org/wso2/carbon/identity/sso/saml/dto/SAMLSSOServiceProviderDTO.java]: use clone of incoming array

File: components/sso-saml/org.wso2.carbon.identity.sso.saml/src/main/java/org/wso2/carbon/identity/sso/saml/processors/IdPInitLogoutRequestProcessor.java [hunk 4]
```diff
@@ -0,0 +1,171 @@
...
+            SSOSessionPersistenceManager ssoSessionPersistenceManager = SSOSessionPersistenceManager
+                    .getPersistenceManager();
+            String sessionIndex = ssoSessionPersistenceManager.getSessionIndexFromTokenId(sessionId);
+            SessionInfoData sessionInfoData = ssoSessionPersistenceManager.getSessionInfo(sessionIndex);
+
+            if (sessionInfoData == null) {
+                log.error(SAMLSSOConstants.Notification.INVALID_SESSION);
+                validationResponseDTO.setValid(false);
+                validationResponseDTO.setLogoutFromAuthFramework(true);
+                ;
```
Comment (hpmtissera) [inline components/sso-saml/org.wso2.carbon.identity.sso.saml/src/main/java/org/wso2/carbon/identity/sso/saml/processors/IdPInitLogoutRequestProcessor.java]: ;

File: components/sso-saml/org.wso2.carbon.identity.sso.saml/src/main/java/org/wso2/carbon/identity/sso/saml/processors/IdPInitLogoutRequestProcessor.java [hunk 5]
```diff
@@ -0,0 +1,171 @@
...
+                    if (!logoutReqIssuer.getIdpInitSLOReturnToURLList().contains(returnTo) && !logoutReqIssuer
+                            .getAssertionConsumerUrlList().contains(returnTo)) {
+                        log.error(SAMLSSOConstants.Notification.INVALID_RETURN_TO_URL);
+                        validationResponseDTO.setValid(false);
+                        return validationResponseDTO;
+                    }
+                    validationResponseDTO.setReturnToURL(returnTo);
+                } else {
+                    validationResponseDTO.setReturnToURL(serverURL.replace(SAMLSSOConstants.SAML_ENDPOINT,
+                                                                                    SAMLSSOConstants
```
Comment (hpmtissera) [inline components/sso-saml/org.wso2.carbon.identity.sso.saml/src/main/java/org/wso2/carbon/identity/sso/saml/processors/IdPInitLogoutRequestProcessor.java]: not formatted properly

File: components/carbon-authenticators/saml2-sso-authenticator/org.wso2.carbon.identity.authenticator.saml2.sso.ui/pom.xml (hunk 1 above)
Comment (thanujalk) [inline components/carbon-authenticators/saml2-sso-authenticator/org.wso2.carbon.identity.authenticator.saml2.sso.ui/pom.xml]: org.wso2.carbon.identity.application.authentication.framework.model.xsd does not have any version and org.wso2.carbon.identity.application.authentication.framework.model.* does not work

**Pull request merged**